### PR TITLE
Issue/2040 configure merge policy

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: docs/configuration/index-config.md [hunk 1]
```diff
@@ -424,13 +424,90 @@ This section describes indexing settings for a given index.
 | ------------- | ------------- | ------------- |
 | `timestamp_field`      | Timestamp field used for sharding documents in splits (1).   | None |
 | `commit_timeout_secs`      | Maximum number of seconds before committing a split since its creation.   | 60 |
-| `split_num_docs_target`      | Maximum number of documents in a split. Note that this is not a hard limit.   | 10_000_000 |
-| `merge_policy.merge_factor`      | Number of splits to merge.   | 10 |
-| `merge_policy.max_merge_factor`      | Maximum number of splits to merge.   | 12 |
+| `split_num_docs_target` | Target of number of docs per split.   | 10_000_000 |
```
Comment (evanxg852000) [inline docs/configuration/index-config.md:427]:
```suggestion
| `split_num_docs_target` | Target number of docs per split.   | 10_000_000 |
```

File: docs/configuration/index-config.md [hunk 2]
```diff
@@ -424,13 +424,90 @@ This section describes indexing settings for a given index.
...
+| `merge_policy` | Describes the strategy used to trigger split merge operations (see [Merge policies] section below). |
```
Comment (evanxg852000) [inline docs/configuration/index-config.md:428]:
```suggestion
| `merge_policy` | Describes the strategy used to trigger split merge operations (see [Merge policies](#merge-policies) section below). |
```

File: docs/configuration/index-config.md [hunk 3]
```diff
@@ -424,13 +424,90 @@ This section describes indexing settings for a given index.
...
+
+
+```yaml
+version: 0
+index_id: "hdfs"
+# ...
+indexing_settings:
+  merge_policy:
+    type: "limit_merge"
+    min_level_num_docs: 100_000
```
Comment (evanxg852000) [inline docs/configuration/index-config.md:482]: Did you mean?
```suggestion
    max_merge_ops: 5
```

File: docs/configuration/index-config.md [hunk 4]
```diff
@@ -424,13 +424,90 @@ This section describes indexing settings for a given index.
...
+
+Quickwit offers three different merge policies, each with their
+own set of parameters.
+
+#### "Stable log" merge policy
+
+The stable log merge policy attempts to minimize write amplification AND keep time-pruning power as high as possible, by merging splits with a similar size, and with a close time span.
+
+Quickwit's default merge policy is a stable log merge policy
+with the following parameter:
```
Comment (boraarslan) [inline docs/configuration/index-config.md:445]:
```suggestion
Quickwit's default merge policy is the `stable_log` merge policy
with the following parameters:
```

File: docs/configuration/index-config.md [hunk 5]
```diff
@@ -424,13 +424,90 @@ This section describes indexing settings for a given index.
...
+| `max_merge_ops`   |  Maximum number of merges that a given split should undergo. | 4 |
+| `merge_factor`      | *(advanced)* Number of splits to merge together in a single merge operation.   | 10 |
+| `max_merge_factor` | *(advanced)* Maximum number of splits that can be merged together in a single merge operation.  | 12 |
+
+#### No merge
+
+The `no_merge` merge policy entirely disables merging.
+This setting is not recommended.
+Merges are necessary to reduce the number of splits,
+and hence improve search performances.
```
Comment (evanxg852000) [inline docs/configuration/index-config.md:500]:
```suggestion
:::caution
This setting is not recommended. Merges are necessary to reduce the number of splits,
and hence improve search performances.
:::
```

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs [hunk 6]
```diff
@@ -19,59 +19,46 @@
...
-/// over-trading write amplification for read amplification here.
-///
-/// The `ConstWriteAmplificationMergePolicy` is very simple. It targets a number
-/// of merges instead, and stops once this number of merges is reached.
-///
-/// Only splits with the same number of merge operations are merged together,
-/// and for a given merge operation, we build split in a greedy way.
-/// After sorting the splits per creation date, we append splits one after the
-/// other until we either reach `max_merge_factor` or we exceed the
-/// targeted` split_num_docs`.
```
Comment (boraarslan) [inline quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:46]: Is this comment irrelevant now?

File: quickwit/quickwit-indexing/src/merge_policy/stable_log_merge_policy.rs [hunk 7]
```diff
@@ -57,25 +58,16 @@ use crate::merge_policy::{splits_short_debug, MergeOperation, MergePolicy};
...
 }
 
 impl Default for StableLogMergePolicy {
     fn default() -> Self {
         StableLogMergePolicy {
-            min_level_num_docs: 100_000,
-            merge_factor: 10,
-            max_merge_factor: 12,
+            config: Default::default(),
             split_num_docs_target: 10_000_000,
```
Comment (evanxg852000) [inline quickwit/quickwit-indexing/src/merge_policy/stable_log_merge_policy.rs:71]: Would be nice to use `IndexingSettings::default_split_num_docs_target()`

**Review by evanxg852000 — APPROVED** [review]
I left some small suggestions.
Question: What's the plan with demuxing as the params (demux_field, demux_enabled) have been deprecated? do we want to separate merge policy from demux?

**Review by fmassot — APPROVED** [review]
LGTM. Just need to fix some docs.

File: quickwit/quickwit-indexing/src/actors/indexing_service.rs [hunk 8]
```diff
@@ -345,7 +346,9 @@ impl IndexingService {
             pipeline_ord: 0,
         };
         let mut index_metadata = self.index_metadata(ctx, &pipeline_id.index_id).await?;
-        index_metadata.indexing_settings.merge_enabled = merge_enabled;
+        if !merge_enabled {
+            index_metadata.indexing_settings.merge_policy = MergePolicyConfig::Nop
```
[the remark this answers is not in the document]
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/actors/indexing_service.rs:350]: agreed. I created and auto assigned myself a ticket to have a look at this.

File: quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs (hunk 6 above)
Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/merge_policy/const_write_amplification.rs:46]: nice catch. it was lost in a refactoring battle. readdingit.

File: docs/configuration/index-config.md (hunk 3 above)
Comment (fulmicoton) [reply docs/configuration/index-config.md:482]: yes thank you

**Pull request merged**